File: bootstrap/legend-pure-next-bootstrap/legend-pure-next-bootstrap-core/legend-pure-next-bootstrap-runtime/src/main/java/org/finos/legend/pure/runtime/PureRuntime.java
// Copyright 2024 Goldman Sachs
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.finos.legend.pure.runtime;

import org.finos.legend.pure.m3.pureLanguage.PureLanguageExtension;
import org.finos.legend.pure.m3.module.CompilationStatistics;
import java.util.Set;
import java.util.Map;
import java.util.LinkedHashSet;
import java.util.LinkedHashMap;
import org.finos.legend.pure.m3.module.Module;
import meta.pure.metamodel.PackageableElement;
import org.finos.legend.pure.m3.module.inMemoryModule.InMemoryModule;
import org.finos.legend.pure.m3.module.ModuleRegistry;
import meta.pure.metamodel.function.FunctionWithParameters;
import org.finos.legend.pure.execution.Execution;
import org.finos.legend.pure.execution.ExecutionContext;
import org.finos.legend.pure.execution.InterpretedExecution;
import org.finos.legend.pure.m3.LanguageExtension;
import org.finos.legend.pure.m3.compilation.Compilation;
import org.finos.legend.pure.m3.compilation.CompilationContext;
import org.finos.legend.pure.m3.module.CompilationResult;
import org.finos.legend.pure.m3.module.MetadataAccess;
import org.finos.legend.pure.m3.module.sourceModule.SourceModule;
import org.finos.legend.pure.next.parser.GrammarExtension;
import org.finos.legend.pure.runtime.compilation.PureCompilation;

import java.util.ArrayList;
import java.util.List;

/**
 * Brings everything together: a registry, the language extensions (parser + compiler
 * per section) and grammars, one {@link Compilation} strategy and one {@link Execution}
 * strategy. Compiling and executing both go through the runtime; natives are configured
 * on the execution strategy.
 *
 * <p>Usage:
 * <pre>{@code
 * PureRuntime runtime = PureRuntime.builder()
 *         .withRegistry(registry)
 *         .withLanguageExtensions(languageExtensions)
 *         .withCompilation(JavaCompilation.builder())          // default: PureCompilation
 *         .withExecution(InterpretedExecution.builder().withNativeExtensions(nativeExtensions))
 *         .build();
 * runtime.setSource("sample", "sample.pure", source);             // an InMemoryModule registered as "sample"
 * CompilationResult compiled = runtime.compile();
 * Object result = runtime.execute(myFunction, "World");
 * }</pre>
 */
public class PureRuntime
{
    private final MetadataAccess registry;
    private final Execution execution;
    private final Compilation compilation;
    /** The extensions the registry's modules are attached with: PureLanguageExtension + the language extensions. */
    private final List<LanguageExtension> moduleExtensions;

    private PureRuntime(MetadataAccess registry,
                        List<LanguageExtension> languageExtensions,
                        List<GrammarExtension> grammarExtensions,
                        Compilation.Builder compilation,
                        Execution.Builder execution)
    {
        this.registry = registry;
        this.moduleExtensions = withPureLanguage(languageExtensions);
        if (registry instanceof ModuleRegistry modules)
        {
            modules.attach(this.moduleExtensions);
        }
        // The compilation executes Pure functions through this runtime (resolved at call
        // time); the execution's natives (compileSource) compile through the compilation.
        this.compilation = compilation.build(new CompilationContext(registry, languageExtensions, this::execute));
        this.execution = execution.build(new ExecutionContext(registry, languageExtensions, grammarExtensions, () -> this.compilation));
    }

    public static Builder builder()
    {
        return new Builder();
    }

    public static class Builder
    {
        private MetadataAccess registry;
        private final List<LanguageExtension> languageExtensions = new ArrayList<>();
        private final List<GrammarExtension> grammarExtensions = new ArrayList<>();
        private Compilation.Builder compilation;
        private Execution.Builder execution;

        public Builder withRegistry(MetadataAccess registry)
        {
            this.registry = registry;
            return this;
        }

        /** Language extensions: the parser and compiler for each extra {@code ###Section}. */
        public Builder withLanguageExtensions(Iterable<? extends LanguageExtension> extensions)
        {
            if (extensions != null)
            {
                extensions.forEach(this.languageExtensions::add);
            }
            return this;
        }

        /**
         * Register grammar extensions consulted by the {@code parseAntlr}
         * native. Each registered extension answers to one
         * {@link GrammarExtension#grammarName()}; duplicates throw at build time.
         * M3 and Top are always available.
         */
        public Builder withGrammarExtensions(Iterable<? extends GrammarExtension> extensions)
        {
            if (extensions != null)
            {
                extensions.forEach(this.grammarExtensions::add);
            }
            return this;
        }

        /** The compilation strategy; defaults to {@link PureCompilation} (compiler.pdb must be loaded to compile). */
        public Builder withCompilation(Compilation.Builder compilation)
        {
            this.compilation = compilation;
            return this;
        }

        /** The execution strategy; defaults to {@link InterpretedExecution} with the built-in natives. */
        public Builder withExecution(Execution.Builder execution)
        {
            this.execution = execution;
            return this;
        }

        public PureRuntime build()
        {
            return new PureRuntime(this.registry, this.languageExtensions, this.grammarExtensions,
                    this.compilation != null ? this.compilation : PureCompilation.builder(),
                    this.execution != null ? this.execution : InterpretedExecution.builder());
        }
    }

    public MetadataAccess registry()
    {
        return this.registry;
    }

    public Compilation compilation()
    {
        return this.compilation;
    }

    public Execution execution()
    {
        return this.execution;
    }

    /** Compile a source module against the registry with this runtime's compilation strategy. */
    public CompilationResult compile(SourceModule module)
    {
        return this.compilation.compile(module);
    }

    /** Add or replace source {@code sourceId} in the registered in-memory module {@code moduleName}. */
    public void setSource(String moduleName, String sourceId, String content)
    {
        inMemoryModule(moduleName).setSource(sourceId, content);
    }

    /** Remove source {@code sourceId} from the registered in-memory module {@code moduleName}. */
    public void removeSource(String moduleName, String sourceId)
    {
        inMemoryModule(moduleName).removeSource(sourceId);
    }

    /**
     * Compile every module whose code lives in the runtime, with this runtime's compilation
     * strategy: first the registered {@link SourceModule}s, in place and in one pass; then, in
     * dependency order, the in-memory modules with code. Each module's elements are replaced by what its sources
     * compile to, so later modules compile against the fresh elements. Stops at the first module
     * that fails; that module keeps its previous elements.
     */
    public CompilationResult compile()
    {
        ModuleRegistry modules = moduleRegistry();
        modules.sortByDependencies();
        List<PackageableElement> compiled = new ArrayList<>();
        Map<String, Set<String>> referencedBy = new LinkedHashMap<>();
        List<CompilationStatistics> statistics = new ArrayList<>();
        if (modules.modules().stream().anyMatch(SourceModule.class::isInstance))
        {
            CompilationResult result = this.compilation.compileSourceModules();
            if (!result.errors().isEmpty())
            {
                return result;
            }
            modules.invalidate();
            compiled.addAll(result.elements());
            statistics.add(result.statistics());
            if (result.referencedBy() != null)
            {
                result.referencedBy().forEach((path, callers) -> referencedBy.computeIfAbsent(path, k -> new LinkedHashSet<>()).addAll(callers));
            }
        }
        for (Module module : List.copyOf(modules.modules()))
        {
            if (!(module instanceof InMemoryModule target) || !target.hasCode())
            {
                continue;
            }
            List<PackageableElement> previous = target.elements();
            target.clearElements();
            modules.invalidate();
            CompilationResult result = this.compilation.compile(
                    new SourceModule(target.name(), target.packagePattern(), target.dependencies(), target.sources()));
            if (!result.errors().isEmpty())
            {
                target.addElements(previous);
                modules.invalidate();
                return new CompilationResult(compiled, result.errors(), result.statistics(), result.referencedBy());
            }
            target.addElements(result.elements());
            target.attach(modules, this.moduleExtensions);
            modules.invalidate();
            compiled.addAll(result.elements());
            statistics.add(result.statistics());
            if (result.referencedBy() != null)
            {
                result.referencedBy().forEach((path, callers) -> referencedBy.computeIfAbsent(path, k -> new LinkedHashSet<>()).addAll(callers));
            }
        }
        return new CompilationResult(compiled, List.of(), CompilationStatistics.combine(statistics), referencedBy);
    }

    private static List<LanguageExtension> withPureLanguage(List<LanguageExtension> languageExtensions)
    {
        List<LanguageExtension> extensions = new ArrayList<>();
        if (languageExtensions.stream().noneMatch(PureLanguageExtension.class::isInstance))
        {
            extensions.add(new PureLanguageExtension());
        }
        extensions.addAll(languageExtensions);
        return List.copyOf(extensions);
    }

    private ModuleRegistry moduleRegistry()
    {
        if (!(this.registry instanceof ModuleRegistry modules))
        {
            throw new IllegalStateException("Modifying and compiling modules needs a ModuleRegistry as the runtime's registry");
        }
        return modules;
    }

    private InMemoryModule inMemoryModule(String moduleName)
    {
        if (!(moduleRegistry().module(moduleName) instanceof InMemoryModule module))
        {
            throw new IllegalArgumentException("No in-memory module '" + moduleName + "' is registered");
        }
        return module;
    }

    /**
     * Execute a compiled function (user-defined, native, or any other
     * {@link FunctionWithParameters}) with the given arguments, in parameter order.
     */
    public Object execute(FunctionWithParameters function, Object... args)
    {
        return this.execution.execute(function, args);
    }

    public void close()
    {
        this.execution.close();
    }
}

File: platforms/javascript/src/core/runtime/PureRuntime.js
// ©2026 JP Morgan Chase & Co. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// PureRuntime for the JavaScript host: provides everything translated code needs from the
// host — metadata access over the registry, the natives, parsing through its language
// extensions, source text — on ONE global, `globalThis.__pureHost`. runtime-lib's
// `__metadataRead`, `__getChild`, `__pureParseTop`, … forward to it. It also executes
// translated functions. JavaScript extras: evalJs (evaluates emitted JS into the shared
// global scope) and the in-memory runtime module (translated globals are its
// function store).

import { metadataHost } from "../compiler/module/pdbModule/marshal.js";
import { NativeRegistry, hostHookFor } from "../execution/natives/NativeRegistry.js";
import { DEFAULT_LANGUAGE_EXTENSIONS } from "./LanguageExtensions.js";
import { translateProgramElements } from "../execution/translate-elements.js";

// The translated compiler-pure entry point, silent overload (generated/compiler.js).
const COMPILE = "meta$pure$compiler$compile_PureFile_MANY__Boolean_1__Boolean_1__CompilationResult_1_";
const asArray = (v) => (v === undefined || v === null ? [] : Array.isArray(v) ? v : [v]);

/** Modules in dependency order (a module after the modules it depends on). */
function dependencyOrder(modules) {
    const byName = new Map(modules.map((m) => [m.name, m]));
    const ordered = [], visited = new Set();
    const visit = (module) => {
        if (visited.has(module.name)) return;
        visited.add(module.name);
        for (const dependency of module.dependencies ?? []) if (byName.has(dependency)) visit(byName.get(dependency));
        ordered.push(module);
    };
    modules.forEach(visit);
    return ordered;
}

// The translated parser-mappings entry point (parser-mappings-bundle.js).
const PARSE_DOCUMENT = "meta$pure$parser$mappings$interpreter$parseDocument_AntlrContext_1__String_1__Boolean_1__Pair_MANY__PureFile_1_";

export class PureRuntime {
    static builder() { return new PureRuntimeBuilder(); }

    #registry;
    #natives;
    #languageExtensions;
    #evalJs;
    #runtimeModule;
    #host = Object.create(null);

    /** Use PureRuntime.builder(). */
    constructor({ registry, natives, languageExtensions, evalJs, runtimeModule, sourceText }) {
        this.#registry = registry;
        this.#natives = natives;
        this.#languageExtensions = languageExtensions;
        this.#evalJs = evalJs;
        this.#runtimeModule = runtimeModule;
        Object.assign(this.#host, metadataHost(registry));
        for (const [signature, factory] of natives.entries()) this.#install(hostHookFor(signature), factory(this));
        // Pure's `parse` (and compileSource's source form) compile to runtime-lib's __pureParseTop.
        this.#install("pureParseTop", (sourceId, content) => this.parse(sourceId, content));
        // Source text of evaluated JavaScript, for Pure stack traces (runtime-lib's call sites).
        if (sourceText) this.#install("hostSourceText", sourceText);
        globalThis.__pureHost = this.#host;
    }

    #install(name, value) {
        this.#host[name] = value;
    }

    get registry() { return this.#registry; }
    get natives() { return this.#natives; }
    get languageExtensions() { return this.#languageExtensions; }
    get evalJs() { return this.#evalJs; }
    get runtimeModule() { return this.#runtimeModule; }
    /** What this runtime provides to translated code (installed as globalThis.__pureHost). */
    get host() { return this.#host; }

    /**
     * Parse Pure source into a PureFile: the Top grammar splits the source into sections, then
     * each section is parsed by the parser a registered language extension contributes for it.
     * Source without a section header is a `###Pure` section.
     */
    parse(sourceId, content) {
        const parseDocument = globalThis[PARSE_DOCUMENT];
        if (typeof globalThis.__parseAntlr !== "function" || typeof parseDocument !== "function") {
            throw new Error("PureRuntime.parse: the parser bundles are not loaded (antlr-bundle.js, parser-mappings-bundle.js)");
        }
        const syntheticHeader = !content.startsWith("###");
        const effective = syntheticHeader ? "###Pure\n" + content : content;
        const document = globalThis.__parseAntlr(effective, "TopParser", sourceId, 0n);
        return parseDocument(document, sourceId, syntheticHeader, this.sectionParsers());
    }

    /** The section parsers of every registered language extension. */
    sectionParsers() {
        return this.#languageExtensions.flatMap((extension) => extension.sectionParsers(this));
    }

    /** Add or replace source `sourceId` in the registered in-memory module `moduleName`. */
    setSource(moduleName, sourceId, content) { this.#sourceModule(moduleName).setSource(sourceId, content); }

    /** Remove source `sourceId` from the registered in-memory module `moduleName`. */
    removeSource(moduleName, sourceId) { this.#sourceModule(moduleName).removeSource(sourceId); }

    #sourceModule(name) {
        const module = this.#registry.module(name);
        if (!module) throw new Error(`no module '${name}' is registered`);
        if (typeof module.setSource !== "function") throw new Error(`module '${name}' does not hold sources`);
        return module;
    }

    /**
     * Compile, in dependency order, the in-memory modules with code: parse their sources through the
     * language extensions, compile them with compiler-pure (translated), then replace the module's
     * elements with the result and evaluate their translated JavaScript — so later modules, and
     * execute, see the fresh elements. Stops at the first module that fails; that module keeps its
     * previous elements. Returns `{ elements, errors }` over the modules compiled.
     */
    compile() {
        const elements = [], errors = [];
        for (const module of dependencyOrder(this.#registry.modules)) {
            if (typeof module.hasCode !== "function" || !module.hasCode()) continue;
            const compile = globalThis[COMPILE];
            if (typeof compile !== "function") throw new Error("PureRuntime.compile: compiler-pure is not loaded (generated/compiler.js)");
            if (typeof this.#evalJs !== "function") throw new Error("PureRuntime.compile: withEvalJs(evalJs) is required to run compiled code");
            let result;
            try {
                result = compile(module.sources().map(({ sourceId, content }) => this.parse(sourceId, content)), false, true);
            } catch (e) {
                errors.push(`${module.name}: ${e && e.message ? e.message : e}`);
                break;
            }
            const moduleErrors = asArray(result.errors).map(String);
            if (moduleErrors.length) {
                errors.push(...moduleErrors);
                break;
            }
            const compiled = asArray(result.elements);
            module.clearElements();
            translateProgramElements(compiled, this.#evalJs, module.name, module);
            module.invalidate?.();
            elements.push(...compiled);
        }
        return { elements, errors };
    }

    /** Execute `fn` — an element from registry.getElement(path), or its path — with `args`. */
    execute(fn, ...args) {
        const path = typeof fn === "string" ? fn : fn?.__purePath;
        if (typeof path !== "string") {
            throw new Error("PureRuntime.execute: expected a function element from registry.getElement(path)");
        }
        // Translated functions are globals any in-memory module resolves; prefer the runtime module.
        const resolver = this.#runtimeModule ?? this.#registry.modules.find((m) => typeof m.resolveFn === "function");
        const impl = resolver?.resolveFn(path, args.length);
        if (typeof impl !== "function") throw new Error(`no translated function for ${path}/${args.length}`);
        return impl(...args);
    }

    /** Uninstall this runtime's host object (when it is still the installed one). */
    close() {
        if (globalThis.__pureHost === this.#host) delete globalThis.__pureHost;
    }
}

export class PureRuntimeBuilder {
    #registry = null;
    #natives = null;
    #languageExtensions = null;
    #evalJs = null;
    #runtimeModule = null;
    #sourceText = null;

    withRegistry(registry) { this.#registry = registry; return this; }
    withNatives(natives) { this.#natives = natives; return this; }
    /** Language extensions (their section parsers); DEFAULT_LANGUAGE_EXTENSIONS when not set. */
    withLanguageExtensions(extensions) { this.#languageExtensions = [...extensions]; return this; }
    withEvalJs(evalJs) { this.#evalJs = evalJs; return this; }
    withRuntimeModule(runtimeModule) { this.#runtimeModule = runtimeModule; return this; }
    /** `fileName -> source text` of evaluated JavaScript, for Pure stack traces. */
    withSourceText(sourceText) { this.#sourceText = sourceText; return this; }

    build() {
        if (!this.#registry) throw new Error("PureRuntime.builder(): withRegistry(registry) is required");
        return new PureRuntime({
            registry: this.#registry,
            natives: this.#natives ?? new NativeRegistry(),
            languageExtensions: this.#languageExtensions ?? DEFAULT_LANGUAGE_EXTENSIONS,
            evalJs: this.#evalJs,
            runtimeModule: this.#runtimeModule,
            sourceText: this.#sourceText,
        });
    }
}
